File: parser.go
package main

import (
	"fmt"
	"strings"
)

type SExp interface {
	String() string
}

type Atom struct {
	value string
}

func (a Atom) String() string {
	return a.value
}

type SList struct {
	children []SExp
}

func (l SList) String() string {
	var sb strings.Builder
	sb.WriteString("(")
	for _, child := range l.children {
		sb.WriteString(child.String())
	}
	sb.WriteString(")")
	return sb.String()
}

func tokenize(input string) []string {
	input = strings.ReplaceAll(input, "(", " ( ")
	input = strings.ReplaceAll(input, ")", " ) ")
	tokens := strings.Fields(input)
	return tokens
}

func parseTokens(tokens []string) (SExp, []string, error) {
	if len(tokens) == 0 {
		return nil, tokens, fmt.Errorf("unexpected end of input")
	}

	token := tokens[0]
	tokens = tokens[1:]
	if token == "(" {
		children := []SExp{}
		for len(tokens) > 0 && tokens[0] != ")" {
			child, rest, err := parseTokens(tokens)
			if err != nil {
				return nil, tokens, err
			}
			children = append(children, child)
			tokens = rest
		}
		if len(tokens) == 0 {
			return nil, tokens, fmt.Errorf("unmatched (")
		}
		tokens = tokens[1:] // Remove ")"
		return SList{children: children}, tokens, nil
	} else if token == ")" {
		return nil, tokens, fmt.Errorf("unmatched )")
	} else {
		return Atom{value: token}, tokens, nil
	}
}

func parse(input string) (SExp, error) {
	tokens := tokenize(input)
	sexp, rest, err := parseTokens(tokens)
	if err != nil {
		return nil, err
	}
	if len(rest) != 0 {
		return nil, fmt.Errorf("invalid s-expression")
	}
	return sexp, nil
}

func main() {
	input := "(+ 1 (* 2 3))"
	result, err := parse(input)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Println(result)
}

File: parser.rs
#[derive(Debug, PartialEq)]
enum SExp {
    Atom(String),
    SList(Vec<SExp>),
}

impl SExp {
    fn to_string(&self) -> String {
        match self {
            SExp::Atom(v) => v.clone(),
            SExp::SList(l) => {
                let mut result = String::from("(");
                for sexp in l {
                    result.push_str(&sexp.to_string());
                }
                result.push_str(")");
                result
            }
        }
    }
}

fn tokenize(input: &str) -> Vec<String> {
    let input = input.replace("(", " ( ").replace(")", " ) ");
    input.split_whitespace().map(String::from).collect()
}

fn parse_tokens(tokens: &mut Vec<String>) -> Result<SExp, String> {
    if tokens.is_empty() {
        return Err("Unexpected end of input".to_string());
    }

    let token = tokens.remove(0);
    match token.as_str() {
        "(" => {
            let mut children = Vec::new();
            while !tokens.is_empty() && tokens[0] != ")" {
                children.push(parse_tokens(tokens)?);
            }
            if tokens.is_empty() {
                return Err("Unmatched (".to_string());
            }
            tokens.remove(0); // Remove ")"
            Ok(SExp::SList(children))
        }
        ")" => Err("Unmatched )".to_string()),
        _ => Ok(SExp::Atom(token)),
    }
}

fn parse(input: &str) -> Result<SExp, String> {
    let mut tokens = tokenize(input);
    let result = parse_tokens(&mut tokens)?;
    if !tokens.is_empty() {
        return Err("Invalid S-expression".to_string());
    }
    Ok(result)
}

fn main() {
    let input = "(+ 1 (* 2 3))";
    match parse(input) {
        Ok(result) => println!("{}", result.to_string()),
        Err(e) => println!("Error: {}", e),
    }
}
